elks/include/arch/stat.h

```diff
@@ -5,11 +5,7 @@
 
 struct stat {
     dev_t	st_dev;
-#ifdef CONFIG_32BIT_INODES
-    __u16	st_ino;
-#else
-    ino_t	st_ino;
-#endif
+    u_ino_t	st_ino;
```

Comment:
Not compliant to standard.

Reply:
It looks like first the stat.h is not correct : ino_t is already & correctly defined in linux/types.h, this conditional in arch/stat.h is useless and erroneous.

Reply:
I'll let you decide what to do about it all... I remain concerned about the fact that the size changes in my comment above, and that affect on applications, that's all. After all, this is the reason why `ls -F` is failing, and nobody knew about it.

Reply:
Yeah, you well spotted the defect, thanks for that, but I think we should cleanup the room first to avoid building our house on wrong basements.

Reply:
> It looks like first the stat.h is not correct : ino_t is already & correctly defined in linux/types.h, this conditional in arch/stat.h is useless and erroneous.

No, the way it is currently coded ensures that `st_ino` is always 16 bits, removing the conditionals would break the ABI. My fix ensures that `st_ino` is always 32 bits.

With no change, `ls -F` can be easily fixed with a `printf` change, but will report incorrect i-node numbers when running large FAT16 or FAT32 disks (in which case the inode number is 'fake' anyways. With this change, the inode number would always be correct, but apparently is not POSIX compliant. Allowing the field member to change size with different kernel compilations would break the ABI and be a larger mistake.

After more thought, perhaps it's easier to just fix `ls -F` using the first method, which is no change to stat.h and a simpler change to `ls`.

Reply:
Hmm... I will have to find the time to look into these changes more closely. :-|

(Perhaps another possibility is to add a new `stat32` or `stat64` syscall, and then try to figure out how to wrap up all the underlying syscalls inside a `stat (`...`)` library routine the userland level (argh).)

Reply:
> (Perhaps another possibility is to add a new `stat32` or `stat64` syscall, and then try to figure out how to wrap up all the underlying syscalls inside a `stat (`...`)` library routine at the userland level (argh).)

That seems like overkill when the only time 32-bit inodes are needed is when reading a FAT32 or large FAT16 disk. Even in those cases, returning a 32-bit inode number has dubious value, as the FAT 'inodes' are generated by the driver, not real inodes.

Reply:
Fixed in a clean manner in commits 12f1995c and 32f25812.